### PR TITLE
Prepare & execute topics in BufferWriteActor

### PR DESCRIPTION
**Pull request by dahbka-lis**
### Changelog entry

Add support for topic operations in BufferWriteActor with enabled OLTP sink.

...

### Changelog category

* Improvement

### Additional information

...

### CODE REVIEW

File: ydb/core/kqp/common/kqp_tx_manager.cpp [hunk 1]
```diff
@@ -235,11 +271,14 @@ class TKqpTransactionManager : public IKqpTransactionManager {
         for (auto& [shardId, shardInfo] : ShardsInfo) {
             if ((shardInfo.Flags & EAction::WRITE)) {
                 ReceivingShards.insert(shardId);
-                if (IsVolatile()) {
-                    SendingShards.insert(shardId);
-                }
-                if (shardInfo.IsOlap) {
-                    sendingColumnShardsSet.insert(shardId);
+                
+                if (!shardInfo.IsTopic) {
```
Comment (nikvas0) [inline ydb/core/kqp/common/kqp_tx_manager.cpp:275]: Don't need it. 
IsVolatile = false in presence of topics.
Olap logic with columnShardArbiter should work fine with topics

File: ydb/core/kqp/executer_actor/kqp_data_executer.cpp [hunk 2]
```diff
@@ -2103,7 +2108,8 @@ class TKqpDataExecuter : public TKqpExecuterBase<TKqpDataExecuter, EExecType::Da
         }
 
         // Single-shard datashard transactions are always immediate
-        ImmediateTx = (datashardTxs.size() + evWriteTxs.size() + Request.TopicOperations.GetSize() + sourceScanPartitionsCount) <= 1
+        auto topicSize = (BufferActorId) ? TxManager->GetTopicOperations().GetSize() : Request.TopicOperations.GetSize();
```
Comment (nikvas0) [inline ydb/core/kqp/executer_actor/kqp_data_executer.cpp:2111]: This is a part of old execution. There is no need to support TxManager here.

File: ydb/core/kqp/executer_actor/kqp_data_executer.cpp [hunk 3]
```diff
@@ -2383,7 +2389,11 @@ class TKqpDataExecuter : public TKqpExecuterBase<TKqpDataExecuter, EExecType::Da
             }
         }
 
-        Request.TopicOperations.BuildTopicTxs(topicTxs);
+        if (BufferActorId) {
```
Comment (nikvas0) [inline ydb/core/kqp/executer_actor/kqp_data_executer.cpp:2392]: This is also part of old execution

File: ydb/core/kqp/executer_actor/kqp_data_executer.cpp [hunk 4]
```diff
@@ -2418,9 +2428,16 @@ class TKqpDataExecuter : public TKqpExecuterBase<TKqpDataExecuter, EExecType::Da
...
-            Request.TopicOperations.HasReadOperations() || Request.TopicOperations.HasWriteOperations())
+        
+        auto hasReadOperations = (BufferActorId)
+            ? TxManager->GetTopicOperations().HasReadOperations()
+            : Request.TopicOperations.HasReadOperations();
+        
+        auto hasWriteOperations = (BufferActorId)
+            ? TxManager->GetTopicOperations().HasWriteOperations()
+            : Request.TopicOperations.HasWriteOperations();
+
```
Comment (nikvas0) [inline ydb/core/kqp/executer_actor/kqp_data_executer.cpp:2439]: Old execution

File: ydb/core/kqp/executer_actor/kqp_data_executer.cpp [hunk 5]
```diff
@@ -2473,12 +2490,16 @@ class TKqpDataExecuter : public TKqpExecuterBase<TKqpDataExecuter, EExecType::Da
                     }
                 }
 
-                if (auto tabletIds = Request.TopicOperations.GetSendingTabletIds()) {
+                if (auto tabletIds = (BufferActorId)
```
Comment (nikvas0) [inline ydb/core/kqp/executer_actor/kqp_data_executer.cpp:2493]: Also don't need it

File: ydb/core/kqp/executer_actor/kqp_data_executer.cpp [hunk 6]
```diff
@@ -2473,12 +2490,16 @@ class TKqpDataExecuter : public TKqpExecuterBase<TKqpDataExecuter, EExecType::Da
...
+                    ? TxManager->GetTopicOperations().GetSendingTabletIds()
+                    : Request.TopicOperations.GetSendingTabletIds()) {
                     sendingShardsSet.insert(tabletIds.begin(), tabletIds.end());
                     receivingShardsSet.insert(tabletIds.begin(), tabletIds.end());
                 }
 
-                if (auto tabletIds = Request.TopicOperations.GetReceivingTabletIds()) {
+                if (auto tabletIds = (BufferActorId)
```
Comment (nikvas0) [inline ydb/core/kqp/executer_actor/kqp_data_executer.cpp:2500]: Also don't need it

File: ydb/core/kqp/executer_actor/kqp_data_executer.cpp [hunk 7]
```diff
@@ -2766,7 +2788,12 @@ class TKqpDataExecuter : public TKqpExecuterBase<TKqpDataExecuter, EExecType::Da
     void ExecuteTopicTabletTransactions(TTopicTabletTxs& topicTxs) {
         YQL_ENSURE(!TxManager);
         TMaybe<ui64> writeId;
-        if (Request.TopicOperations.HasWriteId()) {
+
+        if (BufferActorId && TxManager->GetTopicOperations().HasWriteId()) {
```
Comment (nikvas0) [inline ydb/core/kqp/executer_actor/kqp_data_executer.cpp:2792]: Also don't need it

File: ydb/core/kqp/executer_actor/kqp_data_executer.cpp [hunk 8]
```diff
@@ -2804,7 +2831,9 @@ class TKqpDataExecuter : public TKqpExecuterBase<TKqpDataExecuter, EExecType::Da
             state.DatashardState.ConstructInPlace();
             state.DatashardState->Follower = false;
 
-            state.DatashardState->ShardReadLocks = Request.TopicOperations.TabletHasReadOperations(tabletId);
+            state.DatashardState->ShardReadLocks = (BufferActorId)
```
Comment (nikvas0) [inline ydb/core/kqp/executer_actor/kqp_data_executer.cpp:2834]: Also don't need it

File: ydb/core/kqp/executer_actor/kqp_data_executer.cpp [hunk 9]
```diff
@@ -2742,11 +2763,12 @@ class TKqpDataExecuter : public TKqpExecuterBase<TKqpDataExecuter, EExecType::Da
             ExecuteTopicTabletTransactions(TopicTxs);
         }
 
+        auto topicSize = (BufferActorId) ? TxManager->GetTopicOperations().GetSize() : Request.TopicOperations.GetSize();
```
Comment (nikvas0) [inline ydb/core/kqp/executer_actor/kqp_data_executer.cpp:2766]: Don't need it

File: ydb/core/kqp/runtime/kqp_write_actor.cpp [hunk 10]
```diff
@@ -1618,6 +1623,70 @@ class TKqpBufferWriteActor :public TActorBootstrapped<TKqpBufferWriteActor>, pub
...
+                for (const ui64 sendingShardId : prepareSettings.SendingShards) {
+                    transaction.AddSendingShards(sendingShardId);
+                }
+                for (const ui64 receivingShardId : prepareSettings.ReceivingShards) {
+                    transaction.AddReceivingShards(receivingShardId);
+                }
+            } else {
+                transaction.AddSendingShards(*prepareSettings.ArbiterColumnShard);
+                transaction.AddReceivingShards(*prepareSettings.ArbiterColumnShard);
+                if (prepareSettings.SendingShards.contains(tabletId)) {
```
Comment (nikvas0) [inline ydb/core/kqp/runtime/kqp_write_actor.cpp:1654]: Don't need it if we use ArbiterColumShard

File: ydb/core/kqp/runtime/kqp_write_actor.cpp [hunk 11]
```diff
@@ -1618,6 +1623,70 @@ class TKqpBufferWriteActor :public TActorBootstrapped<TKqpBufferWriteActor>, pub
...
+            transaction.SetImmediate(false);
+
+            ActorIdToProto(SelfId(), ev->Record.MutableSourceActor());
+            ev->Record.MutableData()->Swap(&transaction);
+            ev->Record.SetTxId(*TxId);
+
+            SendTime[tabletId] = TInstant::Now();
+            auto traceId = BufferWriteActor.GetTraceId();
+
+            CA_LOG_D("SendToTopics traceId.verbosity: " << std::to_string(traceId.GetVerbosity()));
```
Comment (nikvas0) [inline ydb/core/kqp/runtime/kqp_write_actor.cpp:1678]: Second log line is enough. Don't need "SendToTopics traceId.verbosity"

File: ydb/core/kqp/common/kqp_tx_manager.h [hunk 12]
```diff
@@ -49,6 +51,10 @@ class IKqpTransactionManager {
     virtual EShardState GetState(ui64 shardId) const = 0;
     virtual void SetState(ui64 shardId, EShardState state) = 0;
 
+    virtual NTopic::TTopicOperations& GetTopicOperations() = 0;
```
Comment (nikvas0) [inline ydb/core/kqp/common/kqp_tx_manager.h:54]: const NTopic::TTopicOperations& GetTopicOperations() const

For changing use SetTopicOperations(...)

**Review by nikvas0 — CHANGES_REQUESTED** [review]

File: ydb/public/sdk/cpp/client/ydb_topic/ut/topic_to_table_ut.cpp [hunk 13]
```diff
@@ -1,7 +1,12 @@
 #include "ut_utils/topic_sdk_test_setup.h"
 
+#include <ydb/core/kqp/ut/common/kqp_ut_common.h>
+#include <ydb/core/testlib/common_helper.h>
+#include <ydb/core/driver_lib/run/run.h>
```
Comment (Gazizonoki) [inline ydb/public/sdk/cpp/client/ydb_topic/ut/topic_to_table_ut.cpp:5]: Мы хотим тесты SDK полностью развязать от серверного кода, можно обойтись без них?

File: ydb/core/kqp/session_actor/kqp_session_actor.cpp [hunk 14]
```diff
@@ -1361,7 +1361,12 @@ class TKqpSessionActor : public TActorBootstrapped<TKqpSessionActor> {
         if (Settings.TableService.GetEnableOltpSink() && !txCtx->TxManager) {
             txCtx->TxManager = CreateKqpTransactionManager();
         }
-        if (Settings.TableService.GetEnableOltpSink() && !txCtx->BufferActorId && txCtx->HasTableWrite) {
+
+        if (Settings.TableService.GetEnableOltpSink()
+            && !txCtx->BufferActorId
+            && (txCtx->HasTableWrite || request.TopicOperations.GetSize() != 0)) {
+            txCtx->TxManager->GetTopicOperations() = std::move(request.TopicOperations);
```
Comment (Alek5andr-Kotov) [inline ydb/core/kqp/session_actor/kqp_session_actor.cpp:1368]: может лучше
```
txCtx->TxManager->SetTopicOperations(std::move(request.TopicOperations));
```
?

File: ydb/public/sdk/cpp/client/ydb_topic/ut/topic_to_table_ut.cpp [hunk 15]
```diff
@@ -1115,6 +1127,20 @@ auto TFixture::GetAvgWriteBytes(const TString& topicName,
...
 }
 
+bool TFixture::GetEnableOltpSink() const
+{
+    return false;
+}
+
+void TFixture::TryReadFromTable(const TString& tablePath,
+                                        NTable::TSession* session,
+                                        NTable::TTransaction* tx)
```
Comment (Alek5andr-Kotov) [inline ydb/public/sdk/cpp/client/ydb_topic/ut/topic_to_table_ut.cpp:1137]: `TSession` и `TTransaction` нужно передавать по ссылке

File: ydb/public/sdk/cpp/client/ydb_topic/ut/topic_to_table_ut.cpp [hunk 16]
```diff
@@ -1115,6 +1127,20 @@ auto TFixture::GetAvgWriteBytes(const TString& topicName,
...
+{
+    TString query = Sprintf(R"(SELECT COUNT(*) FROM `%s`)", tablePath.data());
+    auto result = session->ExecuteDataQuery(query, NTable::TTxControl::Tx(*tx)).GetValueSync();
+    UNIT_ASSERT_C(result.IsSuccess(), result.GetIssues().ToString());
+}
```
Comment (Alek5andr-Kotov) [inline ydb/public/sdk/cpp/client/ydb_topic/ut/topic_to_table_ut.cpp:1142]: Есть метод `TFixture::ExecuteDataQuery`. Он здесь не подошёл?

File: ydb/public/sdk/cpp/client/ydb_topic/ut/topic_to_table_ut.cpp [hunk 17]
```diff
@@ -2486,6 +2512,390 @@ Y_UNIT_TEST_F(WriteToTopic_Demo_48, TFixture)
...
+protected:
+    bool GetEnableOltpSink() const override;
+};
+
+bool TFixtureOltpSink::GetEnableOltpSink() const
+{
+    return true;
+}
+
+Y_UNIT_TEST_F(OltpSink_WriteToTopic_1, TFixtureOltpSink)
```
Comment (Alek5andr-Kotov) [inline ydb/public/sdk/cpp/client/ydb_topic/ut/topic_to_table_ut.cpp:2525]: Дублирует тест WriteToTopic_Demo_7. Можно сделась специальный макрос?

File: ydb/core/kqp/executer_actor/kqp_data_executer.cpp (hunk 2 above)
Comment (dahbka-lis) [reply ydb/core/kqp/executer_actor/kqp_data_executer.cpp:2111]: > This is a part of old execution. There is no need to support TxManager here.

DataExecuter doesn't work well without this condition (use after move or assertion failed for check !TxManager or !BufferActorId with enabled OLTP sink).

**Review by nikvas0 — DISMISSED** [review]

**Review by Gazizonoki — APPROVED** [review]

**Review by Alek5andr-Kotov — APPROVED** [review]

**Review by nikvas0 — APPROVED** [review]

**Pull request merged**